arq/jobs.py

```diff
@@ -155,13 +155,22 @@ async def abort(self, *, timeout: Optional[float] = None, poll_delay: float = 0.
         :param poll_delay: how often to poll redis for the job result
         :return: True if the job aborted properly, False otherwise
         """
-        await self._redis.zadd(abort_jobs_ss, {self.job_id: timestamp_ms()})
-        try:
-            await self.result(timeout=timeout, poll_delay=poll_delay)
-        except asyncio.CancelledError:
-            return True
+        job_info = await self.info()
+        if job_info and job_info.score and job_info.score > timestamp_ms():
+            async with self._redis.pipeline(transaction=True) as tr:
+                tr.zrem(self._queue_name, self.job_id)
+                tr.zadd(self._queue_name, {self.job_id: 1})
+                tr.zadd(abort_jobs_ss, {self.job_id: timestamp_ms()})
+                await tr.execute()
+            raise asyncio.TimeoutError
         else:
-            return False
+            await self._redis.zadd(abort_jobs_ss, {self.job_id: timestamp_ms()})
+            try:
+                await self.result(timeout=timeout, poll_delay=poll_delay)
+            except asyncio.CancelledError:
+                return True
+            else:
+                return False
```

Comment:
I think overall clearer to do something like 

```py
job_info = await self.info()
if job_info and job_info.score and job_info.score > timestamp_ms():
    async with self._redis.pipeline(transaction=True) as tr:
        tr.zrem(self._queue_name, self.job_id)
        tr.zadd(self._queue_name, {self.job_id: 1})
try:
    await self.result(timeout=timeout, poll_delay=poll_delay)
...
```
Just to avoid any race conditions and minimise duplicate code.